File: be-1-nestjs/src/auth/util.handle-tokens.ts
import { verifyAccessToken, VerifyAccessTokenResult } from './util.verify-token';
import { refreshAccessToken, RefreshAccessTokenResult } from './util.refresh-token'
import { UserIdentity } from './auth.types';
import { createIdentity } from './util.identity';

export interface HandleTokensResult {
  readonly isValid: boolean;
  readonly error?: any;
  readonly identity?: UserIdentity
  readonly accessToken?: string
}

interface HandleTokensProps {
  accessToken: string;
  refreshToken: string;
}

export const handleTokens = async ({ accessToken, refreshToken }: HandleTokensProps): Promise<HandleTokensResult> => {
  try {
    if (!accessToken) {
      throw new Error('No access token')
    }

    const verifRes: VerifyAccessTokenResult = await verifyAccessToken(accessToken)
    if (verifRes.isValid) {
      return { isValid: true, identity: createIdentity(verifRes) }
    }

    if (refreshToken) {
      const refreshRes: RefreshAccessTokenResult = await refreshAccessToken(refreshToken)
      const newAccessToken = refreshRes.access_token;
      const verifResNew: VerifyAccessTokenResult = await verifyAccessToken(newAccessToken)

      if (newAccessToken) {
        console.log('...new access token', newAccessToken.substring(0, 8))
        return { isValid: true, identity: createIdentity(verifResNew), accessToken: newAccessToken }
      } else {
        throw new Error('No new access token')
      }
    } else {
      throw new Error('No refresh token')
    }

  } catch (error) {
    console.log("ERROR validateAccessToken ---> ", error.message)
    return { isValid: false, error: error.message }
  }
}

File: be-2-flask/app/auth/util_handle_tokens.py
from typing import TypedDict
from .util_verify_token import verify_access_token, VerifyAccessTokenResult
from .util_refresh_token import refresh_access_token, RefreshAccessTokenResult
from .types import UserIdentity
from .util_verify_token import create_identity


class HandleTokensResult(TypedDict):
    is_valid: bool
    error: str | None
    identity: UserIdentity | None
    access_token: str | None


def handle_tokens(access_token: str | None, refresh_token: str | None) -> HandleTokensResult:
    try:
        if access_token == None:
            raise Exception("No access token provided")
        verif_result: VerifyAccessTokenResult = verify_access_token(access_token)

        if verif_result["is_valid"]:
            return {
                "is_valid": True,
                "error": None,
                "identity": create_identity(verif_result),
                "access_token": None,
            }

        if refresh_token == None:
            raise Exception("No refresh token provided")

        refresh_result: RefreshAccessTokenResult = refresh_access_token(refresh_token)
        access_token_new = refresh_result["access_token"]
        verif_result_new: VerifyAccessTokenResult = verify_access_token(access_token_new)

        if verif_result_new["is_valid"]:
            return {
                "is_valid": True,
                "error": None,
                "identity": create_identity(verif_result_new),
                "access_token": access_token_new,
            }

        raise Exception("Refreshing access token is failed")

    except Exception as e:
        return {"is_valid": False, "error": str(e), "identity": None, "access_token": None}
